### PR TITLE
GEODE-7679 Partitioned Region clear is successful while region is being altered

### PR DESCRIPTION
**Pull request by jchen21**
Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [ ] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionClearWithAlterRegionDUnitTest.java [hunk 1]
```diff
@@ -0,0 +1,424 @@
...
+      assertThat(cacheRule.getCache().getRegion(REGION_NAME).size()).isEqualTo(0);
+    });
+
+    AsyncInvocation asyncInvocation2 = server2.invokeAsync(() -> {
+      alterRegionSetCacheLoader();
+    });
+
+    asyncInvocation1.await();
+    asyncInvocation2.await();
+  }
```
Comment (jinmeiliao) [inline geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionClearWithAlterRegionDUnitTest.java:113]: do you need to add some verification code to verify that region is cleared and altered?

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionClearWithAlterRegionDUnitTest.java [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+      assertThat(cacheRule.getCache().getRegion(REGION_NAME).size()).isEqualTo(0);
+    });
+
+    AsyncInvocation asyncInvocation2 = server2.invokeAsync(() -> {
+      alterRegionSetCacheWriter();
+    });
+
+    asyncInvocation1.await();
+    asyncInvocation2.await();
+  }
```
Comment (jinmeiliao) [inline geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionClearWithAlterRegionDUnitTest.java:132]: same thing, probably some verification code here and in other tests

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionClearWithAlterRegionDUnitTest.java (hunk 1 above)
Comment (jchen21) [reply geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionClearWithAlterRegionDUnitTest.java:113]: It is verified at line 104 and line 322.

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionClearWithAlterRegionDUnitTest.java (hunk 2 above)
Comment (jchen21) [reply geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionClearWithAlterRegionDUnitTest.java:132]: It is verified at line 123 and line 330.

**Review by jinmeiliao — APPROVED** [review]

**Review by kirklund — APPROVED** [review]

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionClearWithAlterRegionDUnitTest.java [hunk 3]
```diff
@@ -0,0 +1,533 @@
...
+      Region region = cacheRule.getCache().getRegion(REGION_NAME);
+      assertThat(region.size()).isEqualTo(NUM_ENTRIES);
+    });
+  }
+
+  @Test
+  public void testClearRegionWhileAddingCacheLoader() throws InterruptedException {
+    initialize();
+
+    AsyncInvocation asyncInvocation1 = server1.invokeAsync(() -> {
```
Comment (gesterzhou) [inline geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionClearWithAlterRegionDUnitTest.java:113]: this should be splited into 2 different call. 
setInstance should be called in invoke(), alterRegionSetCacheLoader() should be called in async.

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionClearWithAlterRegionDUnitTest.java [hunk 4]
```diff
@@ -0,0 +1,533 @@
...
+
+  @Test
+  public void testClearRegionWhileChangingEviction() throws InterruptedException {
+    initialize();
+
+    AsyncInvocation asyncInvocation1 = server1.invokeAsync(() -> {
+      DistributionMessageObserver.setInstance(getDistributionMessageObserver());
+      Region region = cacheRule.getCache().getRegion(REGION_NAME);
+      AttributesMutator attributesMutator = region.getAttributesMutator();
+      getBlackboard().waitForGate(GATE_NAME);
```
Comment (gesterzhou) [inline geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionClearWithAlterRegionDUnitTest.java:173]: (1) why in loader and writer test, you did not use waitForGate?
(2) you can define singalGate in afterProcessedMessage, afterReceivedMessage respectively to test in different cases

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionClearWithAlterRegionDUnitTest.java (hunk 3 above)
Comment (jchen21) [reply geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionClearWithAlterRegionDUnitTest.java:113]: Very good point!

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionClearWithAlterRegionDUnitTest.java (hunk 4 above)
Comment (jchen21) [reply geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionClearWithAlterRegionDUnitTest.java:173]: (1) The `waitForGate` is called in `alterRegionSetCacheLoader` and `alterRegionSetCacheWriter`.
(2) I don't see significant difference between `afterProcessedMessage` and `afterReceivedMessage`. It's true their timings are different. But is the difference significant enough to make the concurrent alter region behavior different? For example, causing some error or failure etc.? Can you elaborate a bit? Maybe we can talk about it.

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionClearWithAlterRegionDUnitTest.java (hunk 3 above)
Comment (jchen21) [reply geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionClearWithAlterRegionDUnitTest.java:115]: Fixed.

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionClearWithAlterRegionDUnitTest.java (hunk 4 above)
Comment (jchen21) [reply geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionClearWithAlterRegionDUnitTest.java:271]: There is no `afterReceivedMessasge` in the `DistributionMessgeObserver` API.

I have added additional tests with test hook that uses `beforeProcessMessage`.

**Review by gesterzhou — APPROVED** [review]

**Pull request merged**